syntax = "proto3";
package shardkv;
message JoinConfigs{
    repeated string config = 1;
}
message JoinRequest{
    map<int32,JoinConfigs> groups = 1;
    int32 requestId = 2;
    int32 clientId = 3;
}

message JoinResponse{
    bool isWrongLeader = 1;
}

message QueryRequest{
    int32 gid = 1;
}

message QueryResponse{
    int32 confignum = 1;
    repeated int32 shards = 2;
    map<int32,JoinConfigs> groups = 3;
}

message LeaveRequest{
    repeated int32 groupIds = 1;
    int32 requestId = 2;
    int32 clientId = 3;
}

message LeaveResponse{
    bool isWrongLeader = 1;
}

//将shardId分片移动到gid下
message MoveRequest{
    int32 shardId = 1;
    int32 gid = 2;
    int32 requestId = 3;
    int32 clientId = 4;
}

message MoveResponse{
    bool isWrongLeader = 1;
}

service ShardKVRpc{
    rpc Join(JoinRequest) returns (JoinResponse);
    rpc Query(QueryRequest) returns (QueryResponse);
    rpc Leave(LeaveRequest) returns (LeaveResponse);
    rpc Move(MoveRequest) returns (MoveResponse);
}